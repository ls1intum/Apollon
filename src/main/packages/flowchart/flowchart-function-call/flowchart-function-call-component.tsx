import React, { FunctionComponent } from 'react';
import { FlowchartComponent } from '../flowchart-element/flowchart-component';
import { FlowchartFunctionCall } from './flowchart-function-call';

export const FlowchartFunctionCallComponent: FunctionComponent<Props> = ({ element, scale }) => (
  <FlowchartComponent element={element} scale={scale}>
    <rect width={10 * scale} height="100%" stroke={element.strokeColor || 'black'} x="0" />
    <rect
      width={element.bounds.width - 20 * scale}
      height="100%"
      stroke={element.strokeColor || 'black'}
      x={10 * scale}
    />
    <rect
      width={10 * scale}
      height="100%"
      stroke={element.strokeColor || 'black'}
      x={element.bounds.width - 10 * scale}
    />
  </FlowchartComponent>
);

export interface Props {
  element: FlowchartFunctionCall;
  scale: number;
}
