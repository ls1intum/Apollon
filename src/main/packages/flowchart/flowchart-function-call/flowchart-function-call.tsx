import { FlowchartElementType } from '..';
import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementType } from '../../uml-element-type';

export class FlowchartFunctionCall extends UMLElement {
  type: UMLElementType = FlowchartElementType.FlowchartFunctionCall;

  render(canvas: ILayer): ILayoutable[] {
    return [this];
  }
}
