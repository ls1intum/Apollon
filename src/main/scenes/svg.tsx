import React, { Component } from 'react';
import { DeepPartial } from 'redux';
import { defaults, Styles } from '../components/theme/styles';
import { Components } from '../packages/components';
import { UMLElementType } from '../packages/uml-element-type';
import { UMLElements } from '../packages/uml-elements';
import { UMLRelationshipType } from '../packages/uml-relationship-type';
import { UMLRelationships } from '../packages/uml-relationships';
import { ILayer } from '../services/layouter/layer';
import { UMLContainer } from '../services/uml-container/uml-container';
import { UMLElement } from '../services/uml-element/uml-element';
import { UMLRelationship } from '../services/uml-relationship/uml-relationship';
import * as Apollon from '../typings';
import { computeBoundingBoxForElements, IBoundary } from '../utils/geometry/boundary';
import { Point } from '../utils/geometry/point';
import { update } from '../utils/update';
import { Style } from './svg-styles';
import { StoreProvider } from '../components/store/model-store';
import { ModelState } from '../components/store/model-state';
import { ThemeProvider } from 'styled-components';

type Props = {
  model: Apollon.UMLModel;
  options?: Apollon.ExportOptions;
  styles?: DeepPartial<Styles>;
};

type State = {
  bounds: IBoundary;
  elements: UMLElement[];
};

const includeChildren = (
  elements: { [id: string]: UMLElement },
  ids: Set<string>,
  include: Set<string>,
): Set<string> => {
  const result = new Set<string>();
  for (const id of ids) {
    const element = elements[id];
    if (!element) continue;

    const children = new Set<string>(UMLContainer.isUMLContainer(element) ? element.ownedElements : []);
    if (include.has(id)) {
      result.add(id);
      include = new Set<string>([...include, ...children]);
    }
    includeChildren(elements, children, include).forEach(result.add, result);
  }
  return result;
};

const excludeChildren = (
  elements: { [id: string]: UMLElement },
  ids: Set<string>,
  exclude: Set<string>,
): Set<string> => {
  const result = new Set<string>();
  for (const id of ids) {
    const element = elements[id];
    if (!element) continue;

    const children = new Set<string>(UMLContainer.isUMLContainer(element) ? element.ownedElements : []);
    if (!exclude.has(id)) {
      result.add(id);
    } else {
      exclude = new Set<string>([...exclude, ...children]);
    }
    excludeChildren(elements, children, exclude).forEach(result.add, result);
  }
  return result;
};

const getInitialState = ({ model, options }: Props): State => {
  const layer: ILayer = {
    layer: document.createElementNS('http://www.w3.org/2000/svg', 'svg'),
    origin: (): Point => new Point(),
  };
  const apollonElements = model.elements;
  const apollonRelationships = model.relationships;

  const deserialize = (apollonElement: Apollon.UMLElement): UMLElement[] => {
    const element = new UMLElements[apollonElement.type]();
    const apollonChildren: Apollon.UMLElement[] = UMLContainer.isUMLContainer(element)
      ? apollonElements.filter((child) => child.owner === apollonElement.id)
      : [];

    element.deserialize(apollonElement, apollonChildren);
    const [root, ...updates] = element.render(layer, []) as UMLElement[];

    return [root, ...updates];
  };

  const elements = apollonElements.reduce<UMLElement[]>((acc, val) => [...acc, ...deserialize(val)], []);

  const relationships = apollonRelationships.map<UMLRelationship>((apollonRelationship) => {
    const relationship = new UMLRelationships[apollonRelationship.type]();
    relationship.deserialize(apollonRelationship);
    return relationship;
  });

  const elementState = [...elements, ...relationships].reduce<{ [id: string]: UMLElement }>(
    (acc, val) => ({ ...acc, [val.id]: val }),
    {},
  );
  const roots = Object.values(elementState).filter((element) => !element.owner);

  let layout = new Set<string>(Object.values(elementState).map((x) => x.id));
  if (options && options.include) {
    layout = includeChildren(
      elementState,
      new Set<string>(roots.map((element) => element.id)),
      new Set<string>(options.include),
    );
  }
  if (options && options.exclude) {
    layout = excludeChildren(
      elementState,
      new Set<string>(roots.map((element) => element.id)),
      new Set<string>(options.exclude),
    );
  }

  const keepOriginalSize = (options && options.keepOriginalSize) || false;

  const bounds = computeBoundingBoxForElements(
    Object.values(elementState).filter((element) => keepOriginalSize || layout.has(element.id)),
  );

  const margin = getMargin(options?.margin);
  bounds.x -= margin.left;
  bounds.y -= margin.top;
  bounds.width += margin.left + margin.right;
  bounds.height += margin.top + margin.bottom;

  const state = Object.values(elementState)
    .filter((element) => layout.has(element.id))
    .map((element) => {
      element.bounds.x -= bounds.x;
      element.bounds.y -= bounds.y;
      return element;
    });

  return {
    elements: state,
    bounds,
  };
};

const getMargin = (
  margin: Apollon.ExportOptions['margin'] = 15,
): { top: number; right: number; bottom: number; left: number } => {
  if (typeof margin === 'number') {
    return { top: margin, right: margin, bottom: margin, left: margin };
  }
  const result = { top: 0, right: 0, bottom: 0, left: 0 };
  return Object.assign(result, margin);
};

export class Svg extends Component<Props, State> {
  state = getInitialState(this.props);

  render() {
    const { bounds, elements } = this.state;
    const theme: Styles = update(defaults, this.props.styles || {});

    // connect exported svg to redux state, so that connected components can retrieve properties from state
    const state = ModelState.fromModel(this.props.model);

    return (
      <StoreProvider initialState={state}>
        <ThemeProvider theme={theme}>
          <svg
            width={bounds.width + 1}
            height={bounds.height + 1}
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
            fillOpacity={0}
            fill="white"
          >
            <defs>
              <style>{(Style[0] as any)({ theme })}</style>
            </defs>
            {elements.map((element, index) => {
              const ElementComponent = Components[element.type as UMLElementType | UMLRelationshipType];
              return (
                <svg
                  {...element.bounds}
                  key={element.id}
                  className={element.name ? element.name.replace(/[<>]/, '') : ''}
                >
                  <ElementComponent key={index} element={element} />
                </svg>
              );
            })}
          </svg>
        </ThemeProvider>
      </StoreProvider>
    );
  }
}
