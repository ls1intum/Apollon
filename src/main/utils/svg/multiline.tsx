import React, { Component, CSSProperties, SVGProps } from 'react';
import { findDOMNode } from 'react-dom';

const defaultProps = Object.freeze({
  x: 0,
  y: 0,
  dx: 0,
  dy: 0,
  angle: 0,
  width: undefined as number | undefined,
  height: undefined as number | undefined,
  lineHeight: 16,
  capHeight: 11,
  scaleToFit: false,
  textAnchor: 'middle' as 'start' | 'middle' | 'end' | 'inherit',
  verticalAnchor: 'middle' as 'start' | 'middle' | 'end',
});

const getInitialState = (props: Props) => {
  const words = props.children ? props.children.toString().split(/\s+/) : [];
  return { wordsByLines: [{ words, width: 0 }] };
};

type Props = { children: string } & SVGProps<SVGTextElement> & typeof defaultProps;

type State = typeof getInitialState;

export class Multiline extends Component<Props, State> {
  static defaultProps = defaultProps;
  state = getInitialState(this.props);

  spaceWidth = 0;
  wordsWithComputedWidth = [] as { word: string; width: number }[];

  componentDidMount() {
    this.updateWordsByLines(this.props, true);
  }

  componentDidUpdate(previousProps: Readonly<Props>) {
    const needCalculate = previousProps.children !== this.props.children || previousProps.style !== this.props.style;
    if (needCalculate) {
      this.updateWordsByLines(this.props, needCalculate);
    }
  }

  calculateWordWidths = (props: Readonly<Props>) => {
    try {
      const words = props.children ? props.children.toString().split(/\s+/) : [];
      const wordsWithComputedWidth = words.map((word) => ({ word, width: this.getStringWidth(word, props.style) }));

      const spaceWidth = this.getStringWidth('\u00A0', props.style);

      return { wordsWithComputedWidth, spaceWidth };
    } catch (e) {
      return null;
    }
  };

  getStringWidth(str: string, style?: CSSProperties): number {
    try {
      // Calculate length of each word to be used to determine number of words per line
      const container = findDOMNode(this);
      if (!container) {
        return 0;
      }

      const text = document.createElementNS('http://www.w3.org/2000/svg', 'tspan');
      container.appendChild(text);

      Object.assign(text.style, style);
      text.textContent = str;
      const width = text.getComputedTextLength();

      container.removeChild(text);

      return width;
    } catch (e) {
      return 0;
    }
  }

  updateWordsByLines(props: Readonly<Props>, needCalculate: boolean) {
    // Only perform calculations if using features that require them (multiline, scaleToFit)
    if (props.width || props.scaleToFit) {
      if (needCalculate) {
        const wordWidths = this.calculateWordWidths(props);

        if (wordWidths) {
          const { wordsWithComputedWidth, spaceWidth } = wordWidths;

          this.wordsWithComputedWidth = wordsWithComputedWidth;
          this.spaceWidth = spaceWidth;
        } else {
          this.updateWordsWithoutCalculate(props);

          return;
        }
      }

      const wordsByLines = this.calculateWordsByLines(this.wordsWithComputedWidth, this.spaceWidth, props.width);
      this.setState({ wordsByLines });
    } else {
      this.updateWordsWithoutCalculate(props);
    }
  }

  updateWordsWithoutCalculate(props: Readonly<Props>) {
    const words = props.children ? props.children.toString().split(/\s+/) : [];
    this.setState({ wordsByLines: [{ words, width: 0 }] });
  }

  calculateWordsByLines(
    wordsWithComputedWidth: { word: string; width: number }[],
    spaceWidth: number,
    lineWidth?: number,
  ): { words: string[]; width: number }[] {
    const { scaleToFit } = this.props;
    return wordsWithComputedWidth.reduce<{ words: string[]; width: number }[]>((result, { word, width }) => {
      const currentLine = result[result.length - 1];

      if (currentLine && (!lineWidth || scaleToFit || currentLine.width + width + spaceWidth < lineWidth)) {
        // Word can be added to an existing line
        currentLine.words.push(word);
        currentLine.width += width + spaceWidth;
      } else {
        // Add first word to line or word is too long to scaleToFit on existing line
        const newLine = { words: [word], width };
        result.push(newLine);
      }

      return result;
    }, []);
  }

  render() {
    const { x, y, dx, dy, textAnchor, verticalAnchor, scaleToFit, angle, lineHeight, capHeight, fill, ...textProps } =
      this.props;
    const { wordsByLines } = this.state;

    const xPosition = x + dx;
    const yPosition = y + dy;

    let startDy: string | number | undefined;
    switch (verticalAnchor) {
      case 'start':
        startDy = capHeight;
        break;
      case 'middle':
        startDy = ((wordsByLines.length - 1) / 2) * -lineHeight + capHeight / 2;
        break;
      default:
        startDy = wordsByLines.length - 1 * -lineHeight;
        break;
    }

    const transforms = [];
    if (scaleToFit && wordsByLines.length) {
      const lineWidth = wordsByLines[0].width;
      const sx = (this.props.width || 0) / lineWidth;
      const sy = sx;
      const originX = xPosition - sx * xPosition;
      const originY = yPosition - sy * yPosition;
      transforms.push(`matrix(${sx}, 0, 0, ${sy}, ${originX}, ${originY})`);
    }
    if (angle) {
      transforms.push(`rotate(${angle}, ${xPosition}, ${yPosition})`);
    }
    if (transforms.length) {
      textProps.transform = transforms.join(' ');
    }

    return (
      <text
        style={fill ? { fill } : {}}
        x={xPosition}
        y={yPosition}
        textAnchor={textAnchor}
        {...textProps}
        pointerEvents="none"
      >
        {wordsByLines.map((line, index) => (
          <tspan x={xPosition} dy={index === 0 ? startDy : lineHeight} key={index}>
            {line.words.join(' ')}
          </tspan>
        ))}
      </text>
    );
  }
}
