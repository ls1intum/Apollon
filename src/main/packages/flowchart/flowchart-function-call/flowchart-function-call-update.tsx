import React, { FunctionComponent } from 'react';
import { FlowchartUpdateComponent, enhance, GeneralProps } from '../flowchart-element/flowchart-update';
import { FlowchartFunctionCall } from './flowchart-function-call';

export const FlowchartFunctionCallUpdateComponent: FunctionComponent<Props> = (props) => {
  return <FlowchartUpdateComponent {...props} />;
};

type OwnProps = {
  element: FlowchartFunctionCall;
};

export type Props = OwnProps & GeneralProps;

export const FlowchartFunctionCallUpdate = enhance(FlowchartFunctionCallUpdateComponent);
