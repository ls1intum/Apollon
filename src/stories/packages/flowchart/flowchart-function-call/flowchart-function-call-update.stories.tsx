import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import { FlowchartFunctionCall } from '../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call';
import {
  FlowchartFunctionCallUpdate,
  Props,
} from '../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call-update';

export default {
  title: 'Packages/Flowchart/Flowchart Function Call Update',
  component: FlowchartFunctionCallUpdate,
} as Meta;

const Template: Story<Props> = (args) => <FlowchartFunctionCallUpdate {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  element: new FlowchartFunctionCall({
    bounds: { x: 0, y: 0, width: 200, height: 100 },
    id: '',
    name: 'Flowchart Function Call',
    owner: null,
  }),
};
