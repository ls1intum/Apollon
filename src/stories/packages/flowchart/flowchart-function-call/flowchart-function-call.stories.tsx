import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import { FlowchartFunctionCall } from '../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call';
import {
  FlowchartFunctionCallComponent,
  Props,
} from '../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call-component';

export default {
  title: 'Packages/Flowchart/Flowchart Function Call',
  component: FlowchartFunctionCallComponent,
  args: { isSvg: true },
} as Meta;

const Template: Story<Props> = (args) => <FlowchartFunctionCallComponent {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  element: new FlowchartFunctionCall({
    bounds: { x: 0, y: 0, width: 200, height: 100 },
    id: '',
    name: 'Flowchart Function Call',
    owner: null,
  }),
};
