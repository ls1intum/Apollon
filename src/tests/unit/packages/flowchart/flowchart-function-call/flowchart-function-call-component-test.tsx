import * as React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { FlowchartFunctionCall } from '../../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call';
import { FlowchartFunctionCallComponent } from '../../../../../main/packages/flowchart/flowchart-function-call/flowchart-function-call-component';
import { Multiline } from '../../../../../main/utils/svg/multiline';
import { CSSProperties } from 'react';

// override getStringWidth, because it uses by jsdom unsupported SVGElement methods
Multiline.prototype.getStringWidth = (str: string, style?: CSSProperties) => {
  return 0;
};

it('render the flowchart-function-call-component', () => {
  const functionCall: FlowchartFunctionCall = new FlowchartFunctionCall({ name: 'TestFunctionCallComponent' });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <FlowchartFunctionCallComponent element={functionCall} />
    </svg>,
  );
  expect(getByText(functionCall.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
