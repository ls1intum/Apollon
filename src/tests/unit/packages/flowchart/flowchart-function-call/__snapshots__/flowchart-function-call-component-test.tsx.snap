// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the flowchart-function-call-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <rect
          class="sc-hKMtZM kFuMUP"
          fill="white"
          height="100%"
          stroke="black"
          width="10"
          x="0"
        />
        <rect
          class="sc-hKMtZM kFuMUP"
          fill="white"
          height="100%"
          stroke="black"
          width="180"
          x="10"
        />
        <rect
          class="sc-hKMtZM kFuMUP"
          fill="white"
          height="100%"
          stroke="black"
          width="10"
          x="190"
        />
        <text
          font-weight="bold"
          height="100"
          pointer-events="none"
          text-anchor="middle"
          width="200"
          x="100"
          y="50"
        >
          <tspan
            dy="5.5"
            x="100"
          >
            TestFunctionCallComponent
          </tspan>
        </text>
      </g>
    </svg>
  </div>
</body>
`;
